server/api/src/rsDataObjRead.cpp

```diff
@@ -55,40 +61,64 @@ applyRuleForPostProcForRead( rsComm_t *rsComm, bytesBuf_t *dataObjReadOutBBuf, c
 }
 
 int
-rsDataObjRead( rsComm_t *rsComm, openedDataObjInp_t *dataObjReadInp,
-               bytesBuf_t *dataObjReadOutBBuf ) {
-    int bytesRead;
-    int l1descInx = dataObjReadInp->l1descInx;
-
-    if ( l1descInx < 2 || l1descInx >= NUM_L1_DESC ) {
-        rodsLog( LOG_NOTICE,
-                 "rsDataObjRead: l1descInx %d out of range",
-                 l1descInx );
+rsDataObjRead(rsComm_t* rsComm,
+              openedDataObjInp_t* dataObjReadInp,
+              bytesBuf_t* dataObjReadOutBBuf)
+{
+    const int l1descInx = dataObjReadInp->l1descInx;
+
+    if (l1descInx < 2 || l1descInx >= NUM_L1_DESC) {
+        rodsLog(LOG_ERROR, "%s: l1descInx %d out of range", __func__, l1descInx);
         return SYS_FILE_DESC_OUT_OF_RANGE;
     }
-    if ( L1desc[l1descInx].inuseFlag != FD_INUSE ) {
+
+    auto& l1desc = L1desc[l1descInx];
+
+    if (l1desc.inuseFlag != FD_INUSE) {
         return BAD_INPUT_DESC_INDEX;
     }
 
-    if ( L1desc[l1descInx].remoteZoneHost != NULL ) {
-        /* cross zone operation */
-        dataObjReadInp->l1descInx = L1desc[l1descInx].remoteL1descInx;
-        bytesRead = rcDataObjRead( L1desc[l1descInx].remoteZoneHost->conn,
-                                   dataObjReadInp, dataObjReadOutBBuf );
+    if (l1desc.remoteZoneHost) {
+        // Cross zone operation.
+        dataObjReadInp->l1descInx = l1desc.remoteL1descInx;
+        const auto bytes_read = rcDataObjRead(l1desc.remoteZoneHost->conn, dataObjReadInp, dataObjReadOutBBuf);
         dataObjReadInp->l1descInx = l1descInx;
+        return bytes_read;
     }
-    else {
-        int i;
-        bytesRead = l3Read( rsComm, l1descInx, dataObjReadInp->len,
-                            dataObjReadOutBBuf );
-        i = applyRuleForPostProcForRead( rsComm, dataObjReadOutBBuf,
-                                         L1desc[l1descInx].dataObjInfo->objPath );
-        if ( i < 0 ) {
-            return i;
+
+    // Do not let the client read past the data size in the catalog if the replica
+    // was opened in read-only mode.
```

Comment:
Might be useful to have the explanatory "This is necessary because..." comment here as well

Reply:
I'm not sure what else to say here. Is there something you feel needs to be added that could help us understand this six months from now?

Reply:
I was just noticing that the text is very similar to that found in the lseek code above, so it might be useful here too:

    // If the replica was opened in read-only mode, then don't allow the client
    // to seek past the data size in the catalog. This is necessary because the
    // size of the replica in storage could be larger than the data size in the
    // catalog.

Reply:
Gotcha. Will add more comments like these.